packages/contracts-bedrock/contracts/test/CrossDomainMessenger.t.sol

```diff
@@ -17,3 +23,100 @@ contract CrossDomainMessenger_BaseGas_Test is Messenger_Initializer {
         L1Messenger.baseGas(hex"ff", _minGasLimit);
     }
 }
+
+// CrossDomainMessenger_RelayMessage_Test tests re-entrency of relayMessage.
+contract CrossDomainMessenger_RelayMessage_Test is Messenger_Initializer {
+    // Storage slot of the l2Sender
+    uint256 constant senderSlotIndex = 50;
+
+    address public fuzzedSender;
+    address public sender;
+
+    address public target;
+
+    // Internal helper function to relay a message and perform assertions.
+    function internalRelay(address innerSender) internal {
+        assertEq(sender, L1Messenger.xDomainMessageSender());
+
+        bytes memory callMessage = abi.encodeWithSelector(
+            CrossDomainMessenger_RelayMessage_Test.relayMessage.selector
+        );
+
+        bytes32 hash = Hashing.hashCrossDomainMessage({
+            _nonce: Encoding.encodeVersionedNonce({ _nonce: 0, _version: 1 }),
+            _sender: innerSender,
+            _target: target,
+            _value: 0,
+            _gasLimit: 0,
+            _data: callMessage
+        });
+
+        vm.expectEmit(true, true, true, true);
+        emit FailedRelayedMessage(hash);
+
+        vm.prank(address(op));
+        L1Messenger.relayMessage({
+            _nonce: Encoding.encodeVersionedNonce({ _nonce: 0, _version: 1 }),
+            _sender: innerSender,
+            _target: target,
+            _value: 0,
+            _minGasLimit: 0,
+            _message: callMessage
+        });
+
+        assertTrue(L1Messenger.failedMessages(hash));
+        assertFalse(L1Messenger.successfulMessages(hash));
+        assertEq(sender, L1Messenger.xDomainMessageSender());
+    }
+
+    // relayMessage is called by the L1 Cross Domain Messenger.
+    function relayMessage() external payable {
+        for (uint256 i = 0; i < 10; i++) {
+            address innerSender;
+            unchecked {
+                innerSender = address(uint160(uint256(uint160(fuzzedSender)) + i));
+            }
+            internalRelay(innerSender);
+        }
+    }
+
+    function testFuzz_relayMessageReenter_succeeds(address _sender) external {
+        vm.assume(_sender != Predeploys.L2_CROSS_DOMAIN_MESSENGER);
+        fuzzedSender = _sender;
+
+        target = address(this);
+        sender = Predeploys.L2_CROSS_DOMAIN_MESSENGER;
+        bytes memory callMessage = abi.encodeWithSelector(
+            CrossDomainMessenger_RelayMessage_Test.relayMessage.selector
+        );
+
+        vm.expectCall(target, callMessage);
+
+        bytes32 hash = Hashing.hashCrossDomainMessage({
+            _nonce: Encoding.encodeVersionedNonce({ _nonce: 0, _version: 1 }),
+            _sender: sender,
+            _target: target,
+            _value: 0,
+            _gasLimit: 0,
```

Comment:
Is it worth fuzzing this value too?

Reply:
I think fuzzing the gaslimit here makes sense as well as the gaslimit in the top level call, just in case there is a weird unnoticed interaction. There is also the possibility of making the `relayMessage` call with a certain amount of gas